### PR TITLE
Consolidate event handling into Listener type.

### PR DESCRIPTION
**Pull request by Laimiux**
## What
- Added `Listener` type (maybe should call it `Callback` instead). This replaces `() -> Unit` and `(Event) -> Unit`
- Renamed `TransitionCallbackWrapper` to `TransitionDispatcher`.

### CODE REVIEW

File: formula/src/main/java/com/instacart/formula/BoundStream.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package com.instacart.formula
+
+class BoundStream<Message>(
```
Comment (Laimiux) [inline formula/src/main/java/com/instacart/formula/BoundStream.kt:3]: Renamed from `Update` to `BoundStream`

File: formula/src/main/java/com/instacart/formula/Listener.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.instacart.formula
+
+/**
+ * Formula uses Listener callbacks to pass events.
+ */
+fun interface Listener<Event> : (Event) -> Unit {
```
Comment (Laimiux) [inline formula/src/main/java/com/instacart/formula/Listener.kt:6]: Still debating what would be the right name for this.

File: formula/src/main/java/com/instacart/formula/TransitionContext.kt [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package com.instacart.formula
+
+interface TransitionContext {
```
Comment (Laimiux) [inline formula/src/main/java/com/instacart/formula/TransitionContext.kt:3]: Not yet sure about moving this out.

File: formula/src/main/java/com/instacart/formula/Update.kt [hunk 4]
```diff
@@ -1,41 +1,13 @@
...
-
-        return true
-    }
+/**
+ * An update function takes an [Event] and calculates the [Transition] that should happen. A
+ * transition can contain a new [state][State] object which would trigger a [Formula.evaluate]
+ * and/or [Effects] that should be performed. You can return [Transition.None] if nothing
+ * should happen.
+ */
+fun interface Update<State, Event> {
```
Comment (Laimiux) [inline formula/src/main/java/com/instacart/formula/Update.kt:9]: I'm wondering how to better unify this with the `Transition` type. Maybe this should be changed to 
```kotlin
fun interface Transition<State, Event> {
    sealed class Result {
       /**
         * Stateful transition.
         *
         * @param state New state
         * @param effects Optional effects such as parent callbacks, logging, db writes,
         * network requests, etc.
         */
        data class Stateful<State>(val state: State, override val effects: Effects? = null) : Result<State>()

        /**
         * Only effects are emitted as part of this transition
         *
         * @param effects Effects such as parent callbacks, logging, db writes, network requests, etc.
         */
        data class OnlyEffects(override val effects: Effects) : Result<Nothing>()

        /**
         * Nothing happens in this transition.
         */
        object None : Result<Nothing>() {
            override val effects: Effects? = null
        }
   }

   fun TransitionContext.create(event: Event): Result<State>
}
```

File: formula/src/main/java/com/instacart/formula/internal/TransitionDispatcher.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package com.instacart.formula.internal
+
+import com.instacart.formula.Transition
+import com.instacart.formula.Update
+
+@PublishedApi
+internal class TransitionDispatcher<State>(
```
Comment (Laimiux) [inline formula/src/main/java/com/instacart/formula/internal/TransitionDispatcher.kt:7]: This was called `TransitionCallbackWrapper` previously

File: formula/src/main/java/com/instacart/formula/internal/TransitionDispatcher.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        if (!terminated && transitionId.hasTransitioned()) {
+            // We have already transitioned, this should not happen.
+            throw IllegalStateException("Transition already happened. This is using old transition callback: $transition.")
+        }
+
+        handleTransition(transition)
+    }
+
+    // TODO: naming UpdateType as transition seems weird. We need to resolve these two types.
+    fun <Event> dispatch(transition: Update<State, Event>, event: Event) {
```
Comment (Laimiux) [inline formula/src/main/java/com/instacart/formula/internal/TransitionDispatcher.kt:32]: TODO: fix this once we resolve naming for `Update`

File: formula/src/main/java/com/instacart/formula/internal/FormulaManagerImpl.kt [hunk 7]
```diff
@@ -175,14 +175,14 @@ internal class FormulaManagerImpl<Input, State, Output>(
...
-        frame?.transitionCallbackWrapper?.terminated = true
-        callbacks.disableAll()
+        frame?.transitionDispatcher?.terminated = true
         children?.forEachValue { it.markAsTerminated() }
     }
 
     override fun performTerminationSideEffects() {
         children?.forEachValue { it.performTerminationSideEffects() }
         updateManager.terminate()
+        callbacks.disableAll()
```
Comment (Laimiux) [inline formula/src/main/java/com/instacart/formula/internal/FormulaManagerImpl.kt:185]: The reason I'm moving this from `markAsTerminated` to `performTerminationSideEffects` is to allow termination side-effects to execute.

File: formula/src/main/java/com/instacart/formula/BoundStream.kt (hunk 1 above)
Comment (Laimiux) [reply formula/src/main/java/com/instacart/formula/BoundStream.kt:3]: Moving this into #246

**Comment by carrotkite** [comment]
### JaCoCo Code Coverage 79.53% :white_check_mark:
| Class | Covered | Meta | Status |
|:---|:---:|:---:|:---:|
| `com/instacart/formula/internal/FormulaManagerImpl` | 88% | 0% | :white_check_mark: |
| `com/instacart/formula/internal/ScopedCallbacks` | 86% | 0% | :white_check_mark: |
| `com/instacart/formula/internal/Frame` | 92% | 0% | :white_check_mark: |
| `com/instacart/formula/internal/CallbackImpl` | 75% | 0% | :white_check_mark: |
| `com/instacart/formula/internal/Callbacks` | 80% | 0% | :white_check_mark: |
| `com/instacart/formula/internal/UpdateManager` | 94% | 0% | :white_check_mark: |
| `com/instacart/formula/internal/FormulaContextImpl` | 50% | 0% | :white_check_mark: |
| `com/instacart/formula/FormulaContext` | 47% | 0% | :white_check_mark: |
| `com/instacart/formula/BoundStream` | 75% | 0% | :white_check_mark: |
| `com/instacart/formula/Evaluation` | 100% | 0% | :white_check_mark: |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: formula/src/main/java/com/instacart/formula/internal/CallbackImpl.kt [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+import com.instacart.formula.Listener
+import com.instacart.formula.Update
+import java.lang.IllegalStateException
+
+/**
+ * Note: this class is not a data class because equality is based on instance and not [key].
+ */
+@PublishedApi
+internal class CallbackImpl<State, Event>(internal var key: Any) : Listener<Event> {
```
Comment (Laimiux) [inline formula/src/main/java/com/instacart/formula/internal/CallbackImpl.kt:11]: TODO: this should be renamed to `ListenerImpl` or `Listener` should be renamed to `Callback`

File: formula/src/main/java/com/instacart/formula/FormulaContext.kt [hunk 9]
```diff
@@ -221,17 +187,14 @@ abstract class FormulaContext<State> internal constructor(
...
-            crossinline transition: Transition.Factory.(Message) -> Transition<State>
+            transition: Update<State, Message>,
         ): BoundStream<Message> {
-            val callback: (Message) -> Unit = {
-                val value = transition(Transition.Factory, it)
-                transitionCallback(value)
-            }
-
+            val key = JoinedKey(stream.key(), transition::class)
+            val callback = formulaContext.onEvent(key, transition)
```
Comment (Laimiux) [inline formula/src/main/java/com/instacart/formula/FormulaContext.kt:195]: This now delegates to `onEvent` to create the `Stream` listener/callback.

File: formula/src/main/java/com/instacart/formula/internal/FormulaManagerImpl.kt (hunk 7 above)
Comment (Laimiux) [reply formula/src/main/java/com/instacart/formula/internal/FormulaManagerImpl.kt:185]: This was needed for https://github.com/instacart/formula/pull/245/files#r719839477

File: formula/src/main/java/com/instacart/formula/FormulaContext.kt [hunk 10]
```diff
@@ -2,120 +2,88 @@ package com.instacart.formula
...
-    }
-
-    /**
-     * Creates a callback that takes a [UIEvent] and performs a [Transition].
+     * Creates a callback that takes an [Event] and performs a [Transition].
      *
-     * It uses inlined callback anonymous class for type.
+     * It uses [transition] type as key.
      */
-    inline fun <UIEvent> onEvent(
```
Comment (alexanderbezverhni) [inline formula/src/main/java/com/instacart/formula/FormulaContext.kt:50]: Any reason we migrated off `inline` for all functions here?

File: formula/src/main/java/com/instacart/formula/Listener.kt [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+
+}
+
+/**
+ * Utility function used to specific
```
Comment (alexanderbezverhni) [inline formula/src/main/java/com/instacart/formula/Listener.kt:11]: INcomplete comment?

File: formula/src/main/java/com/instacart/formula/internal/CallbackImpl.kt (hunk 8 above)
Comment (alexanderbezverhni) [reply formula/src/main/java/com/instacart/formula/internal/CallbackImpl.kt:11]: Maybe add this as a comment to this class?

File: formula/src/test/java/com/instacart/formula/subjects/EventCallbackFormula.kt [hunk 12]
```diff
@@ -17,7 +17,7 @@ class EventCallbackFormula : Formula<Unit, String, EventCallbackFormula.Output>
         return Evaluation(
             output = Output(
                 state = state,
-                changeState = context.onEvent { newState ->
+                changeState = context.onEvent<String> { newState ->
```
Comment (alexanderbezverhni) [inline formula/src/test/java/com/instacart/formula/subjects/EventCallbackFormula.kt:20]: This actually adds up to the code readability.

**Review by alexanderbezverhni — APPROVED** [review]
LGTM 👍🏼

File: formula/src/main/java/com/instacart/formula/Listener.kt (hunk 2 above)
Comment (johnwilde) [reply formula/src/main/java/com/instacart/formula/Listener.kt:6]: I prefer `Listener`

File: formula/src/main/java/com/instacart/formula/Update.kt (hunk 4 above)
Comment (johnwilde) [reply formula/src/main/java/com/instacart/formula/Update.kt:9]: Yeah, having an `Update` that produces a `Transition` is a bit confusing, would be better to not have that extra name I think.

**Review by johnwilde — APPROVED** [review]

File: formula/src/main/java/com/instacart/formula/Listener.kt (hunk 2 above)
Comment (mattkranzler5) [reply formula/src/main/java/com/instacart/formula/Listener.kt:6]: Another option would be `Handler`. I don't have a strong preference, just throwing it out there to see what other people think.

**Review by mattkranzler5 — APPROVED** [review]
Looks good, I like the changes.

File: formula/src/main/java/com/instacart/formula/Listener.kt (hunk 2 above)
Comment (Laimiux) [reply formula/src/main/java/com/instacart/formula/Listener.kt:6]: Yeah, I've also thought about `Handler` though I didn't like its associations with `android.Handler`

File: formula/src/main/java/com/instacart/formula/Update.kt (hunk 4 above)
Comment (Laimiux) [reply formula/src/main/java/com/instacart/formula/Update.kt:9]: Does `Update` and `Update.Result` or `Transition` and `Transition.Result` seem better? On one hand, `Update` corresponds to `MVU` naming, however many folks like the name `Transition` (including me).

File: formula/src/main/java/com/instacart/formula/internal/CallbackImpl.kt (hunk 8 above)
Comment (Laimiux) [reply formula/src/main/java/com/instacart/formula/internal/CallbackImpl.kt:11]: Most likely, I'm going to rename this before merging the PR.

File: formula/src/main/java/com/instacart/formula/FormulaContext.kt (hunk 10 above)
Comment (Laimiux) [reply formula/src/main/java/com/instacart/formula/FormulaContext.kt:50]: Previously, we were creating another function inside of here and cross-inlining the `transition` function to save an extra object allocation. I've restructured this, so inline isn't needed anymore.

File: formula/src/main/java/com/instacart/formula/Update.kt (hunk 4 above)
Comment (Laimiux) [reply formula/src/main/java/com/instacart/formula/Update.kt]: I'm moving this change out of this PR.

File: formula/src/main/java/com/instacart/formula/Listener.kt (hunk 11 above)
Comment (Laimiux) [reply formula/src/main/java/com/instacart/formula/Listener.kt]: Updated

File: formula/src/test/java/com/instacart/formula/subjects/EventCallbackFormula.kt (hunk 12 above)
Comment (Laimiux) [reply formula/src/test/java/com/instacart/formula/subjects/EventCallbackFormula.kt:20]: Do you mean improves code readability?

**Comment by Laimiux** [comment]
Hey folks, thank you for your initial review 💯 

To reduce the scope of this PR, I've moved some changes such as `TransitionContext` and `Update` out (I want to fully flesh these out before merging). In this PR, I focused on the `Listener` change and update all the docs. Please take a quick look again to see if anything doesn't make sense cc @johnwilde @mattkranzler5 @alexanderbezverhni

File: formula/src/test/java/com/instacart/formula/subjects/KeyUsingListFormula.kt [hunk 13]
```diff
@@ -17,7 +18,7 @@ class KeyUsingListFormula :
 
     data class ItemRenderModel(
         val item: String,
-        val onDeleteSelected: () -> Unit
+        val onDeleteSelected: Listener<Unit>
```
Comment (jasonostrander) [inline formula/src/test/java/com/instacart/formula/subjects/KeyUsingListFormula.kt:21]: Does this change mean we need to type our callbacks as `Listener` in RenderModels, or is this change specific to these formulas?

**Review by jasonostrander — APPROVED** [review]
LGTM, just one question.

File: formula/src/test/java/com/instacart/formula/subjects/KeyUsingListFormula.kt (hunk 13 above)
Comment (Laimiux) [reply formula/src/test/java/com/instacart/formula/subjects/KeyUsingListFormula.kt:21]: When using `onEvent` you will need to type to `Listener` type. For now, there still exists `callback` and `eventCallback` which support old style.

**Review by Jawnnypoo — APPROVED** [review]

**Pull request merged**